packages/langium/test/validation/document-validator.test.ts

```diff
@@ -110,3 +111,142 @@ describe('Generic `AstNode` validation applies to all nodes', () => {
     });
 
 });
+
+describe('Register Before/AfterDocument logic for validations with state', () => {
+
+    const grammar = `grammar NestedNamedNodes
+
+    entry Model:
+        (elements+=NamedNode)*;
+
+    NamedNode:
+        name=ID '{' (children+=NamedNode)* '}';
+
+    hidden terminal WS: /\\s+/;
+    terminal ID: /[_a-zA-Z][\\w_]*/;
+    `;
+
+    let validate: (input: string) => Promise<ValidationResult<AstNode>>;
+    let services: LangiumServices;
+
+    function getFQN(node: AstNode): string {
+        if ('name' in node && node.$container) {
+            const parentName = getFQN(node.$container);
+            if (parentName) {
+                return parentName + '.' + node.name;
+            }
+            return '' + node.name;
+        }
+        return '';
+    }
+
+    function rememberNamedNode(child: AstNode, fqnMap: Map<string, AstNode[]>): void {
+        const fqn = getFQN(child);
+        let list = fqnMap.get(fqn);
+        if (!list) {
+            list = [];
+            fqnMap.set(fqn, list);
+        }
+        list.push(child);
+    }
+
+    function registerValidationForRootElement() {
+        const validationChecks: ValidationChecks<object> = {
+            // Note that there are no generated types like Model and NamedElement here
+            AstNode: (node, accept) => {
```

Comment:
Even though there are no generated types for the in-memory test language, it's possible to write the corresponding types manually in order to improve the readability and type safety of the tests.

Reply:
Done, it improves the readability a lot, thanks!